### PR TITLE
Improve Scheduler Compatibility Test

### PR DESCRIPTION
**Pull request by ahmad-diaa**
**What type of PR is this?**
/kind cleanup

/sig scheduling
/assign @alculquicondor 

**What this PR does / why we need it**:
- Update compatibility_test to check for policy in the instantiated scheduler instead of checking for the decoded policy
- Moving users of `factory.NewConfigFactory` to `scheduler.New`

**Which issue(s) this PR fixes**:

Part of #71859

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: pkg/scheduler/api/compatibility/compatibility_test.go [hunk 1]
```diff
@@ -1218,34 +1224,48 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
...
 			StatusCode:   500,
-			ResponseBody: "",
+			ResponseBody: string(body),
 			T:            t,
 		}
 		server := httptest.NewServer(&handler)
 		defer server.Close()
 		client := clientset.NewForConfigOrDie(&restclient.Config{Host: server.URL, ContentConfig: restclient.ContentConfig{GroupVersion: &schema.GroupVersion{Group: "", Version: "v1"}}})
+		algorithmSrc := schedulerconfig.SchedulerAlgorithmSource{
+			Policy: &schedulerconfig.SchedulerPolicySource{},
```
Comment (ahg-g) [inline pkg/scheduler/api/compatibility/compatibility_test.go]: shouldn't this specify config map as the algorithm source? The tests should fail without that I believe.

Comment (ahmad-diaa) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: That's because this test is using FakeHandler (line 1236) which was updated to return the ConfigMap. Depending on the server to get the config map won't work in this case (because of the FakeHandler).

Comment (ahg-g) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: I don't get it, scheduler.New below will try to setup the scheduler using ```algorithmSrc```, which seems to me half initialized: it specifies Policy (vs algorithm provider), but does not specify the source of Policy (ConfigMap vs File). I am not sure how this is related to how ConfigMap is served.

File: pkg/scheduler/api/compatibility/compatibility_test.go [hunk 2]
```diff
@@ -1218,34 +1224,48 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
...
+			ObjectMeta: metav1.ObjectMeta{Namespace: metav1.NamespaceSystem, Name: "scheduler-custom-policy-config"},
+			Data:       map[string]string{schedulerconfig.SchedulerPolicyConfigMapKey: tc.JSON},
+		}
+		policyConfigMap.APIVersion = "v1"
+		body, err := json.Marshal(policyConfigMap)
+		if err != nil {
+			t.Errorf("Error creating configMap: %+v", err)
+		}
 		handler := utiltesting.FakeHandler{
 			StatusCode:   500,
```
Comment (alculquicondor) [inline pkg/scheduler/api/compatibility/compatibility_test.go]: 200?

File: pkg/scheduler/api/compatibility/compatibility_test.go [hunk 3]
```diff
@@ -1218,34 +1224,48 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
 		if !reflect.DeepEqual(policy, tc.ExpectedPolicy) {
 			t.Errorf("%s: Expected:\n\t%#v\nGot:\n\t%#v", v, tc.ExpectedPolicy, policy)
 		}
-
+		policyConfigMap := v1.ConfigMap{
```
Comment (alculquicondor) [inline pkg/scheduler/api/compatibility/compatibility_test.go]: I don't think we need to convert `tc.JSON` lines above to a policy anymore. Verifying `ExpectedPolicy` seems out of scope for this unit test.
Verifying the parsing is a good test for `Codec` in `pkg/scheduler/api/latest`, not this.
Instead, here we should verifying that `scheduler.Algorithm` corresponds to the policy.

File: pkg/scheduler/api/compatibility/compatibility_test.go (hunk 1 above)
Comment (ahmad-diaa) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: Aha, I got the point now. We should have an empty ConfigMap inside the Policy as well .. Am I correct?

Comment (ahg-g) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: you need to populate SchedulerPolicyConfigMapSource with the name and namespace of the configmap

Comment (ahg-g) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: I am surprised this is passing the tests as it is...

Comment (alculquicondor) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: Me, not so much. This test is just testing the parsing of the JSON into a policy, instead of the effects in the scheduler. This is wrong, of course.

Comment (ahmad-diaa) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: @alculquicondor but the scheduler did not return any error. empty policy is ok in scheduler creation, nothing checks for it.

Comment (alculquicondor) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: Right, but that means that we are not using the policy that we want to use :)

File: pkg/scheduler/api/compatibility/compatibility_test.go (hunk 3 above)
Comment (alculquicondor) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: @ahg-g, any thoughts?

File: pkg/scheduler/api/compatibility/compatibility_test.go (hunk 1 above)
Comment (ahmad-diaa) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: sure, just wanted to point out a reason why the test passed.

Comment (ahg-g) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: is there a way to check that the policy the scheduler is using is the expected one?

File: pkg/scheduler/api/compatibility/compatibility_test.go (hunk 3 above)
Comment (ahg-g) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: +1 if we can do that, that is conformance.

File: pkg/scheduler/api/compatibility/compatibility_test.go [hunk 4]
```diff
@@ -1218,34 +1224,48 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
...
-			PvInformer:                     informerFactory.Core().V1().PersistentVolumes(),
-			PvcInformer:                    informerFactory.Core().V1().PersistentVolumeClaims(),
-			ReplicationControllerInformer:  informerFactory.Core().V1().ReplicationControllers(),
-			ReplicaSetInformer:             informerFactory.Apps().V1().ReplicaSets(),
-			StatefulSetInformer:            informerFactory.Apps().V1().StatefulSets(),
-			ServiceInformer:                informerFactory.Core().V1().Services(),
-			PdbInformer:                    informerFactory.Policy().V1beta1().PodDisruptionBudgets(),
-			StorageClassInformer:           informerFactory.Storage().V1().StorageClasses(),
-			CSINodeInformer:                informerFactory.Storage().V1beta1().CSINodes(),
-			HardPodAffinitySymmetricWeight: v1.DefaultHardPodAffinitySymmetricWeight,
```
Comment (ingvagabund) [inline pkg/scheduler/api/compatibility/compatibility_test.go]: +1 for removing `HardPodAffinitySymmetricWeight`, `DisablePreemption` and `PercentageOfNodesToScore`. Function `scheduler.New` defaults to `v1.DefaultHardPodAffinitySymmetricWeight`, `false` and `schedulerapi.DefaultPercentageOfNodesToScore` when those settings are not passed through `opts ...func(o *schedulerOptions)`.

File: pkg/scheduler/api/compatibility/compatibility_test.go (hunk 1 above)
Comment (ingvagabund) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: > Me, not so much. This test is just testing the parsing of the JSON into a policy, instead of the effects in the scheduler. This is wrong, of course.

@alculquicondor one would have to run the scheduler and schedule at least one pod to validate the policy can produce a scheduler capable of scheduling. Though it will still not answer the question if the policy is really usable. There are (or should be) other tests validating scheduler's capabilities given a list of policies.

> is there a way to check that the policy the scheduler is using is the expected one?

Are you asking for (new line 1224):

```go
if !reflect.DeepEqual(policy, tc.ExpectedPolicy) {
	t.Errorf("%s: Expected:\n\t%#v\nGot:\n\t%#v", v, tc.ExpectedPolicy, policy)
}
```
?

Comment (ingvagabund) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: > I am surprised this is passing the tests as it is...

@ahg-g unfortunate constellation. `pkg/api/scheduler/api.validation.ValidatePolicy` considers empty `schedulerapi.Policy{}` as valid (everything falls down to defaults). Also:

```go
case source.Policy != nil:
	// Create the config from a user specified policy source.
	policy := &schedulerapi.Policy{}
	switch {
	...
	case source.Policy.ConfigMap != nil:
		if err := initPolicyFromConfigMap(client, source.Policy.ConfigMap, policy); err != nil {
			return nil, err
		}
	}
```

the inner case block does not check if `source.Policy.ConfigMap.Namespace`, resp. `source.Policy.ConfigMap.Name` is set since in relies on the clientset to tell if it exists or not.

Instead, we can use fake clientset when initializing the client:
```go
client := fake.NewSimpleClientset(&policyConfigMap)
```

which will provided the expected error message as the fake client will not find CM with an empty name.

Comment (ingvagabund) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: Surprise, surprise. In case there is only a single object added into the fake client, setting both namespace and name  to an empty string "will" return the object. One might temporarily create another CM:
```
client := fake.NewSimpleClientset(
	&policyConfigMap,,
	&v1.ConfigMap{ObjectMeta: metav1.ObjectMeta{Name: "test"}},
)
```

Upstream issue reported here: https://github.com/kubernetes/client-go/issues/685

Comment (alculquicondor) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: I think we could validate that the policy was applied correctly by calling `scheduler.Algorithm.Predicates()` and `scheduler.Algorithm.Prioritizers()`.

File: pkg/scheduler/api/compatibility/compatibility_test.go [hunk 5]
```diff
@@ -1218,34 +1224,48 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
...
+		}
 		handler := utiltesting.FakeHandler{
 			StatusCode:   500,
-			ResponseBody: "",
+			ResponseBody: string(body),
 			T:            t,
 		}
 		server := httptest.NewServer(&handler)
 		defer server.Close()
 		client := clientset.NewForConfigOrDie(&restclient.Config{Host: server.URL, ContentConfig: restclient.ContentConfig{GroupVersion: &schema.GroupVersion{Group: "", Version: "v1"}}})
```
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: +1

File: pkg/scheduler/api/compatibility/compatibility_test.go (hunk 1 above)
Comment (ahmad-diaa) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: @ingvagabund I believe, this wouldn't be a problem in our case, as the test is creating a single policyConfigMap. Am I correct?

Comment (ahmad-diaa) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: >I think we could validate that the policy was applied correctly by calling scheduler.Algorithm.Predicates() and scheduler.Algorithm.Prioritizers()

How would we compare those to priorities (they are of different types)? should we only compare names?

Comment (alculquicondor) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: Today, this test compares the parsing of the json into an expected policy. What we should do is remove that expected policy and have expected predicates and expected prioritizers.
FYI, in go, prefer the word `want` instead of `expected` for the variables' names.

File: pkg/scheduler/api/compatibility/compatibility_test.go [hunk 6]
```diff
@@ -1218,34 +1219,42 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
...
-
-		handler := utiltesting.FakeHandler{
-			StatusCode:   500,
-			ResponseBody: "",
-			T:            t,
+		policyConfigMap := v1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{Namespace: metav1.NamespaceSystem, Name: "scheduler-custom-policy-config"},
+			Data:       map[string]string{schedulerconfig.SchedulerPolicyConfigMapKey: tc.JSON},
+		}
+		policyConfigMap.APIVersion = "v1"
```
Comment (ingvagabund) [inline pkg/scheduler/api/compatibility/compatibility_test.go]: No need to set it.

File: pkg/scheduler/api/compatibility/compatibility_test.go (hunk 1 above)
Comment (ahmad-diaa) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: @alculquicondor what about extenders? ExtenderConfigs are not exposed by `ScheduleAlgorithm`.

Comment (alculquicondor) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: They seem to be part of Algorithm as well (just not exposed)
https://github.com/kubernetes/kubernetes/blob/a56e86a405a414f83e5bb1d4e84e37d80af543b7/pkg/scheduler/factory/factory.go#L416

We can simply add the interface.

File: pkg/scheduler/api/compatibility/compatibility_test.go [hunk 7]
```diff
@@ -1209,52 +1206,80 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
...
 			continue
 		}
-	}
-
-	if !seenPredicates.HasAll(registeredPredicates.List()...) {
-		t.Errorf("Registered predicates are missing from compatibility test (add to test stanza for version currently in development): %#v", registeredPredicates.Difference(seenPredicates).List())
-	}
-	if !seenPriorities.HasAll(registeredPriorities.List()...) {
-		t.Errorf("Registered priorities are missing from compatibility test (add to test stanza for version currently in development): %#v", registeredPriorities.Difference(seenPriorities).List())
+		schedPredicates := sets.NewString()
```
Comment (ahmad-diaa) [inline pkg/scheduler/api/compatibility/compatibility_test.go]: Not sure if that is the correct way to compare predicates and prioritizers. Would like to hear others recommendations for this.

File: pkg/scheduler/api/compatibility/compatibility_test.go [hunk 8]
```diff
@@ -1209,52 +1206,80 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
...
+		}
+		if !schedPredicates.Equal(wantPredicates) {
+			t.Errorf("Expected predicates %v, got %v", wantPredicates, schedPredicates)
+		}
+		if !schedPrioritizers.Equal(wantPrioritizers) {
+			t.Errorf("Expected prioritizers %v, got %v", wantPrioritizers, schedPrioritizers)
+		}
+		var wantAlgoExtenders []algorithm.SchedulerExtender
+		schedExtenders := sched.Algorithm.Extenders()
+		for _, e := range tc.ExpectedPolicy.ExtenderConfigs {
```
Comment (ahmad-diaa) [inline pkg/scheduler/api/compatibility/compatibility_test.go]: @alculquicondor  Couldn't figure out the right way to compare extenders. The following did not work. Any thoughts?

Comment (alculquicondor) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: from here: https://github.com/kubernetes/kubernetes/blob/a0c93b10c9173476d284ac5a7cd8569bc91dedff/pkg/scheduler/factory/factory.go#L347

it looks like all extenders should be of the same type `HTTPExtender` you can have `wantExtenders []*HttpExtender` that you can compare.

Comment (ahmad-diaa) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: `core.NewHTTPExtender` returns an algorithm.SchedulerExtender. Why would we use HTTPExtender then?

Comment (alculquicondor) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: The point is that all of them are of the same type. And if you want to compare their configuration, that might be the only option.

File: pkg/scheduler/core/extender.go [hunk 9]
```diff
@@ -114,6 +115,36 @@ func NewHTTPExtender(config *schedulerapi.ExtenderConfig) (algorithm.SchedulerEx
 	}, nil
 }
 
+// Equal is used to check if two extenders are equal
+// ignoring the client field, exported for testing
+func Equal(extender1, extender2 *HTTPExtender) bool {
```
Comment (ahmad-diaa) [inline pkg/scheduler/core/extender.go]: Is there a cleaner way to do this comparison? Like, looping over the fields and checking them (with the possibility to skip certain fields)?

File: pkg/scheduler/api/compatibility/compatibility_test.go (hunk 7 above)
Comment (alculquicondor) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: I would just get rid of `expectedPolicy` altogether, and have `wantPredicates: sets.NewString(...)`. Same for prioritizers. The way you are obtaining the names from the Predicates map LGTM.

File: pkg/scheduler/api/compatibility/compatibility_test.go [hunk 10]
```diff
@@ -1254,16 +1239,47 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
...
 			[]kubeschedulerconfig.PluginConfig{},
-		); err != nil {
+		)
+		if err != nil {
 			t.Errorf("%s: Error constructing: %v", v, err)
 			continue
 		}
-	}
-
-	if !seenPredicates.HasAll(registeredPredicates.List()...) {
```
Comment (alculquicondor) [inline pkg/scheduler/api/compatibility/compatibility_test.go]: I feel like this could still be there (it doesn't depend on the policy).

File: pkg/scheduler/api/compatibility/compatibility_test.go [hunk 11]
```diff
@@ -1254,16 +1239,47 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
...
+		schedExtenders := sched.Algorithm.Extenders()
+		var wantExtenders []*core.HTTPExtender
+		for _, e := range tc.ExpectedPolicy.ExtenderConfigs {
+			extender, err := core.NewHTTPExtender(&e)
+			if err != nil {
+				t.Errorf("Error transforming extender: %+v", e)
+			}
+			wantExtenders = append(wantExtenders, extender.(*core.HTTPExtender))
+		}
+		for idx := range schedExtenders {
```
Comment (alculquicondor) [inline pkg/scheduler/api/compatibility/compatibility_test.go]: `i`

File: pkg/scheduler/api/compatibility/compatibility_test.go [hunk 12]
```diff
@@ -1254,16 +1239,47 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
...
+			if !core.Equal(wantExtenders[idx], schedExtenders[idx].(*core.HTTPExtender)) {
+				t.Errorf("Expected extenders %+v, got %+v", wantExtenders[idx], schedExtenders[idx])
```
Comment (alculquicondor) [inline pkg/scheduler/api/compatibility/compatibility_test.go]: `"Want extender #%d %+v, got %+v"`

File: pkg/scheduler/core/extender.go [hunk 13]
```diff
@@ -114,6 +115,36 @@ func NewHTTPExtender(config *schedulerapi.ExtenderConfig) (algorithm.SchedulerEx
...
+	if extender1.bindVerb != extender2.bindVerb {
+		return false
+	}
+	if extender1.weight != extender2.weight {
+		return false
+	}
+	if extender1.nodeCacheCapable != extender2.nodeCacheCapable {
+		return false
+	}
+	if !reflect.DeepEqual(extender1.managedResources, extender2.managedResources) {
```
Comment (alculquicondor) [inline pkg/scheduler/core/extender.go]: prefer `e1.managedResources.Equal(e2.managedResources)`

File: pkg/scheduler/core/extender.go (hunk 9 above)
Comment (alculquicondor) [reply pkg/scheduler/core/extender.go]: probably with `reflect.Value`, but I think this is good enough.
Prefer shorter variable names, like `e1`

File: pkg/scheduler/core/extender.go [hunk 14]
```diff
@@ -114,6 +115,36 @@ func NewHTTPExtender(config *schedulerapi.ExtenderConfig) (algorithm.SchedulerEx
...
+		return false
+	}
+	if extender1.weight != extender2.weight {
+		return false
+	}
+	return true
```
Comment (alculquicondor) [inline pkg/scheduler/core/extender.go]: Did you miss `ignorable`? prefer to use the same order for the comparisons as the struct definition.

Comment (ahmad-diaa) [reply pkg/scheduler/core/extender.go]: They have the same order, just missed this one (a copy paste error :man_facepalming:)

**Comment by ahmad-diaa** [comment]
>I would just get rid of expectedPolicy altogether, and have wantPredicates: sets.NewString(...). Same for prioritizers.

Can we get rid of expectedPolicy but keep wantPredicates as `[]schedulerapi.PredicatePolicy` to avoid having a huge number of changes in the PR which do not seem necessary to me? or is it ok to have such large number of changes?

File: pkg/scheduler/api/compatibility/compatibility_test.go (hunk 7 above)
Comment (ahmad-diaa) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: Can we get rid of expectedPolicy but keep wantPredicates as []schedulerapi.PredicatePolicy to avoid having a huge number of changes in the PR which do not seem necessary to me? or is it ok to have such large number of changes?

File: pkg/scheduler/api/compatibility/compatibility_test.go (hunk 10 above)
Comment (ahmad-diaa) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: From what I understand, It's only testing the policy decoding as well since seenPredicates are extracted from the policy directly.

Comment (alculquicondor) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: I mean, the check can still be there, but using the set that is calculated from `sched.Algorithm`

File: pkg/scheduler/api/compatibility/compatibility_test.go (hunk 7 above)
Comment (alculquicondor) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: If you want you can make a parent PR to change just that, but I think it's ok to have it in this PR.
Just update the description indicating that we are fixing the test to check for the fields in the instantiated scheduler instead of the decoding.

File: pkg/scheduler/api/compatibility/compatibility_test.go [hunk 15]
```diff
@@ -1252,13 +1257,15 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
 		wantPredicates.Insert("CheckNodeCondition") // mandatory predicate
 		for _, p := range tc.ExpectedPolicy.Predicates {
 			wantPredicates.Insert(p.Name)
+			seenPredicates.Insert(p.Name)
```
Comment (alculquicondor) [inline pkg/scheduler/api/compatibility/compatibility_test.go]: We have this information in `schedPredicates` already.

File: pkg/scheduler/api/compatibility/compatibility_test.go [hunk 16]
```diff
@@ -1276,10 +1283,17 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
 			}
 			wantExtenders = append(wantExtenders, extender.(*core.HTTPExtender))
 		}
-		for idx := range schedExtenders {
-			if !core.Equal(wantExtenders[idx], schedExtenders[idx].(*core.HTTPExtender)) {
-				t.Errorf("Expected extenders %+v, got %+v", wantExtenders[idx], schedExtenders[idx])
+		for i := range schedExtenders {
+			if !core.Equal(wantExtenders[i], schedExtenders[i].(*core.HTTPExtender)) {
+				t.Errorf("Want extender %+v, got %+v", wantExtenders[i], schedExtenders[i])
```
Comment (alculquicondor) [inline pkg/scheduler/api/compatibility/compatibility_test.go]: add the index to the error message. Also, got before want: `"Got extender #%d %+v, want %+v"`

File: pkg/scheduler/core/extender.go [hunk 17]
```diff
@@ -181,7 +180,7 @@ func (h *HTTPExtender) ProcessPreemption(
 		// If extender has cached node info, pass NodeNameToMetaVictims in args.
 		nodeNameToMetaVictims := convertToNodeNameToMetaVictims(nodeToVictims)
 		args = &schedulerapi.ExtenderPreemptionArgs{
-			Pod:                   pod,
```
Comment (alculquicondor) [inline pkg/scheduler/core/extender.go]: Revert this line. The linter shouldn't be doing this.

File: pkg/scheduler/api/compatibility/compatibility_test.go (hunk 15 above)
Comment (ahmad-diaa) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: `schedPredicates` are cleared after each test, `seenPredicates` aren't

Comment (alculquicondor) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: Uhm.... that sounds wrong. Can you think of a reason why is that?

Comment (alculquicondor) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: Uhm... this is weird. It looks like it is testing the test. i.e. making sure that the tests cover all the cases.
In any case, I think its more clear if `seenPredicates` is obtained `seenPredicates.Union(schedPredicates)` in each test.

File: pkg/scheduler/api/compatibility/compatibility_test.go [hunk 18]
```diff
@@ -1209,46 +1209,85 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
 			t.Errorf("%s: Error decoding: %v", v, err)
 			continue
 		}
-		for _, predicate := range policy.Predicates {
-			seenPredicates.Insert(predicate.Name)
-		}
-		for _, priority := range policy.Priorities {
-			seenPriorities.Insert(priority.Name)
+		policyConfigMap := v1.ConfigMap{
```
Comment (alculquicondor) [inline pkg/scheduler/api/compatibility/compatibility_test.go]: Let's wrap each case with `t.Run`, instead of doing `Printf`

File: pkg/scheduler/api/compatibility/compatibility_test.go (hunk 15 above)
Comment (ahmad-diaa) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: sg, done!

File: pkg/scheduler/api/compatibility/compatibility_test.go [hunk 19]
```diff
@@ -63,22 +63,21 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
...
-				},
-				Priorities: []schedulerapi.PriorityPolicy{
-					{Name: "LeastRequestedPriority", Weight: 1},
-					{Name: "ServiceSpreadingPriority", Weight: 2},
-					{Name: "TestServiceAntiAffinity", Weight: 3, Argument: &schedulerapi.PriorityArgument{ServiceAntiAffinity: &schedulerapi.ServiceAntiAffinity{Label: "zone"}}},
-					{Name: "TestLabelPreference", Weight: 4, Argument: &schedulerapi.PriorityArgument{LabelPreference: &schedulerapi.LabelPreference{Label: "bar", Presence: true}}},
-				},
-			},
+			wantPredicates: sets.NewString(
+				"CheckNodeCondition", //mandatory predicate
```
Comment (ahg-g) [inline pkg/scheduler/api/compatibility/compatibility_test.go]: why is this added to the compatibility test? it was added only after 1.9: https://github.com/kubernetes/kubernetes/blob/8993fbc543c18e73668793b5d5e234c0a136735c/CHANGELOG-1.9.md

**Review by ahg-g — COMMENTED** [review]
One nit. thanks alot for improving the compatibility test.

File: pkg/scheduler/api/compatibility/compatibility_test.go (hunk 19 above)
Comment (ahmad-diaa) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: that's because the scheduler will add it to the predicates, so if we don't add it there schedPredicates and wantPredicates won't match.

File: pkg/scheduler/api/compatibility/compatibility_test.go [hunk 20]
```diff
@@ -1127,128 +1079,134 @@ func TestCompatibility_v1_Scheduler(t *testing.T) {
...
+			client := fake.NewSimpleClientset(&policyConfigMap)
+			algorithmSrc := schedulerconfig.SchedulerAlgorithmSource{
+				Policy: &schedulerconfig.SchedulerPolicySource{
+					ConfigMap: &kubeschedulerconfig.SchedulerPolicyConfigMapSource{
+						Namespace: policyConfigMap.Namespace,
+						Name:      policyConfigMap.Name,
+					},
+				},
+			}
+			informerFactory := informers.NewSharedInformerFactory(client, 0)
```
Comment (ahg-g) [inline pkg/scheduler/api/compatibility/compatibility_test.go]: mandatoryPredicates := sets.NewString("CheckNodeCondition"), and take the union of mandatory and want when comparing with schedPredicates.

note that we have to keep "CheckNodeCondition" in the wantPredicates set of the releases that actually introduced it.

Comment (ahmad-diaa) [reply pkg/scheduler/api/compatibility/compatibility_test.go]: That's definitely better. Resolved.

**Comment by ahg-g** [comment]
Thanks, please squash the commits

**Comment by alculquicondor** [comment]
Great work @ahmad-diaa

**Comment by ahmad-diaa** [comment]
@alculquicondor Thank you for you constant guidance!
